### PR TITLE
GH677 Allow pd.RangeIndex to be initialized with range

### LINKED ISSUE #677: `pd.RangeIndex` should accept `range` as the first argument
**Describe the bug**
Both according to the docs and the implementation it is valid to pass a python `range` object as the first argument to `pd.RangeIndex`

**To Reproduce**
```python
import pandas as pd

pd.RangeIndex(range(5))  # mypy: arg-type – Argument 1 to "RangeIndex" has incompatible type "range"
```

**Please complete the following information:**
 - OS: Linux (OpenSUSE Tumbleweed)
 - Python 3.10
 - mypy 1.0.1
 - pandas-stubs 2.0.1.230501

### PR DESCRIPTION
**Pull request by loicdiridollou**
- [x] Closes #677 
- [x] Tests added: Please use `assert_type()` to assert the type of any return value

### CODE REVIEW

File: pandas-stubs/core/indexes/range.pyi
Comment (loicdiridollou) [inline pandas-stubs/core/indexes/range.pyi:25]: Also quick note here, the pandas implementation does not have a `__init__` method so we may want to remove that from pandas-stubs, @Dr-Irv let me know if you think it is good to drop.

Comment (loicdiridollou) [reply pandas-stubs/core/indexes/range.pyi:25]: https://github.com/pandas-dev/pandas/blob/2419343bfea5dba678146139ca9663d831c47b22/pandas/core/indexes/range.py#L150-L158

**Comment by loicdiridollou** [comment]
\pandas_nightly

File: pandas-stubs/core/indexes/range.pyi
Comment (twoertwein) [reply pandas-stubs/core/indexes/range.pyi:25]: You can drop __init__.

**Review by twoertwein — APPROVED** [review]
Happy to merge or wait if you want to remove `__init__` first

**Comment by loicdiridollou** [comment]
> Happy to merge or wait if you want to remove `__init__` first

I deleted the `__init__` method, just waiting for CI to finish but it is looking ready from a code perspective.

**Comment by twoertwein** [comment]
Thanks @loicdiridollou !

**Comment by Dr-Irv** [comment]
FYI, I think we should have kept `__init__()` and deleted `__new__()`.  In the stubs, we only need to have `__new__()` if we want to have different overloads that produce different results.  Otherwise `__init__()` is sufficient.

Admittedly, we're inconsistent in the stubs about that. There are times we use `__new__()` instead of `__init__()`

**Comment by twoertwein** [comment]
> Admittedly, we're inconsistent in the stubs about that. There are times we use `__new__()` instead of `__init__()`

If we want to become more stubtest compatible (aligning pandas and pandas-stubs), we should use which ever constructor pandas uses, at least when possible.

**Pull request merged**
